### PR TITLE
Add end-to-end (system) tests

### PR DESCRIPTION
**Pull request by renchap**
This is an early work in progress and I need to tidy up some things, but I would like to have some feedback on it.

With this PR, we can now write end-to-end tests (system tests in rspec/rails lingo) that will run in a real browser (headless chrome by default).

It allows testing the full Mastodon stack, including the front end code and the streaming server.

The intent is not to have full coverage with those tests, but cover the main workflows and features to prevent critical regressions.

There are 2 impactful changes here:
- the streaming server needs to be started and correctly configured for the test app to work. I implemented a simple runner that will start it in the background when a system test will be ran, and stop it afterwards
- the streaming server needs database access, but by default rails wraps every test in a transaction that is rolled back after the test, preventing the streaming server to see any data. I disabled this for the system tests and used `DatabaseCleaner` to wipe the tables between each system test.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -115,6 +115,10 @@ group :test do
 
   # Browser integration testing
   gem 'capybara', '~> 3.39'
+  gem 'selenium-webdriver'
+
+  # Used to reset the database between system tests
+  gem 'database_cleaner-active_record'
```
Comment (ClearlyClaire) [inline Gemfile:121]: Could this be moved to a specific dependency group?

File: config/environments/test.rb [hunk 2]
```diff
@@ -14,6 +14,11 @@
 
   config.assets.digest = false
 
+  config.public_file_server.enabled = true
+  config.public_file_server.headers = {
+    'Cache-Control' => "public, max-age=#{1.hour.to_i}",
+  }
```
Comment (ClearlyClaire) [inline config/environments/test.rb:20]: What is it used for?

File: config/initializers/content_security_policy.rb [hunk 3]
```diff
@@ -35,6 +35,9 @@ def host_to_url(str)
 
     p.connect_src :self, :data, :blob, assets_host, media_host, Rails.configuration.x.streaming_api_base_url, *webpacker_urls
     p.script_src  :self, :unsafe_inline, :unsafe_eval, assets_host
+  elsif Rails.env.test?
+    p.connect_src :self, :data, :blob, assets_host, media_host, Rails.configuration.x.streaming_api_base_url
+    p.script_src  :self, assets_host, :unsafe_eval, "'wasm-unsafe-eval'"
```
Comment (ClearlyClaire) [inline config/initializers/content_security_policy.rb:40]: Why is `unsafe-eval` required here?

**Review by ClearlyClaire — COMMENTED** [review]
Being able to do automatic end-to-end tests would definitely be a good thing, although I think this should not run systematically on the CI given the additional cost, and be optional for development given the additional requirements. This should also have minimal impacts on the existing testsuite.

File: config/environments/test.rb (hunk 2 above)
Comment (ClearlyClaire) [reply config/environments/test.rb:20]: Right, though we use our own middleware for this: https://github.com/mastodon/mastodon/blob/9e245d147bcb2c72cc552ff8c276a1c34e2f686d/config/application.rb#L186-L189

I notice that it's not enabled by default in the test environment though.

File: config/initializers/content_security_policy.rb (hunk 3 above)
Comment (renchap) [reply config/initializers/content_security_policy.rb:40]: I need to re-check, but IIRC this is needed by Selenium as it injects JS into the page to drive the tests

File: config/environments/test.rb (hunk 2 above)
Comment (renchap) [reply config/environments/test.rb:20]: Good catch, I completely missed this!

Do you see any issue with enabling the middleware in tests as well? Or should I try to find a way to have the environment variable set when Puma is launched for the system tests?

Comment (ClearlyClaire) [reply config/environments/test.rb:20]: I can't see any issue in enabling it for tests as well.

**Comment by renchap** [comment]
> I think this should not run systematically on the CI given the additional cost, and be optional for development given the additional requirements. This should also have minimal impacts on the existing testsuite.

The cost is not this high for the moment, the 2 sample tests I added in this PR take 3 seconds to run (+ 5 seconds for starting the various parts).

I think I will add a default tag filter in `rspec` to not run the system tests when using `rails test`, and run them with `rails test:system` (this is the default for Rails app).

For CI, they should be ran, but this can probably be done only with the latest supported Ruby version, and in a step running concurrently with the existing steps. I dont expect this to make the CI significantly longer.

**Comment by nschonni** [comment]
> Being able to do automatic end-to-end tests would definitely be a good thing, although I think this should not run systematically on the CI given the additional cost, and be optional for development given the additional requirements. This should also have minimal impacts on the existing testsuite.

Here is an old example where a special label was used `create-preview` to trigger a particular CI job. That way, the maintainer can decide if they want to run the end-to-end tests, rather than always running them

**Comment by renchap** [comment]
`rails spec:system` runs the system tests, but I am not sure how to ignore them in `rails spec`.

This will probably require overriding this rake task to add `exclude_pattern: 'spec/system/**/*`, but I am unsure how to do it properly.

For reference:
- [the `spec:*` definitions](https://github.com/rspec/rspec-rails/blob/main/lib/rspec/rails/tasks/rspec.rake) (`rspec-rails`)
- [the `spec` task definition](https://github.com/rspec/rspec-core/blob/main/lib/rspec/core/rake_task.rb) (`rspec-core`)

Any idea on the best way to do this?

**Comment by renchap** [comment]
PR status:
- Automatically starting the streaming server before the first system spec is launched works fine, and it is shutdown when the test suite ends
- Specs use an isolated DB environment thanks to `database_cleaner`, seems to work fine as well
- `rails spec` does not launch system specs, but `rails spec:system` does
- System specs are launched in Github Actions, concurrently with the normal specs, in their own runner

TODO:
- I needed to set `LOCAL_DOMAIN` to `localhost` so the system tests work correctly. This need to be global, as the fixtures / DB use this value. I replaced the previous ngrok domain with `localhost`, but 1 non-system spec is failing and I dont know why
- I do not know rspec very well. I wrote `spec/system/new_statuses_spec.rb` as a sample spec, but I am not sure this is idiomatic. Especially because I am re-using `ProfileStories`. I would prefer is someone more confident with rspec can determine the best way to have system tests helpers done correctly.

Pinging @mjankowski as you may have some expertise on this?

**Comment by mjankowski** [comment]
> Pinging @mjankowski as you may have some expertise on this?

Sure!

> * Automatically starting the streaming server before the first system spec is launched works fine, and it is shutdown when the test suite ends
> * Specs use an isolated DB environment thanks to `database_cleaner`, seems to work fine as well
> * `rails spec` does not launch system specs, but `rails spec:system` does
> * System specs are launched in Github Actions, concurrently with the normal specs, in their own runner
> * You need a local install of Chrome + `chromedriver` installed for the tests to run. They should also work on Firefox with `geckodriver`, but I did not add a way to switch browsers

All sounds good so far.

> * I needed to set `LOCAL_DOMAIN` to `localhost` so the system tests work correctly. This need to be global, as the fixtures / DB use this value. I replaced the previous ngrok domain with `localhost`, but 1 non-system spec is failing and I dont know why

Out of curiosity, what was the driver of needing this change? (ie, which part could not be configured to use the existing hostname?) i haven't looked at the node/streaming portion of this PR, but I know in the rails/rack/rspec/capybara world you sometimes need to set this in multiple places config-wise, and/or sometimes need to set the hostname per-example if you need to change it from the default. I'm curious what thing was not working under previous name.

> * I do not know rspec very well. I wrote `spec/system/new_statuses_spec.rb` as a sample spec, but I am not sure this is idiomatic. Especially because I am re-using `ProfileStories`. I would prefer is someone more confident with rspec can determine the best way to have system tests helpers done correctly.

Two part answer here -- first, the actual changes here are fine. Aside from style nitpicks, I think the approach you took of "have the first full end to end spec just be the most basic case of posting a status and nothing more complicated" is a good starting place.

Second, roadmap-wise, I was planning on doing something like:

- Continue to flesh out the C0 coverage across all the controllers/request specs -- I have a few open PRs on this front, and we're not quite to 100%, but we're close and I think the critical paths are tested at this point.
- Migrate the current feature specs to become system specs, including the ProfileStories module. This is almost entirely a rename/move operation and will just update the configuration to use modern rails/rspec terminology and file locations for things.
- Migrate the controller specs for the API specifically to be request specs -- danielmbrasil has done a bunch of these and I assume has more in the works. I think it's viable that close to 100% of the API coverage is in request specs and there are few if any contoller specs left for API area.
- Migrate existing controller specs to be system specs driving the browser (many of the ones for the settings/admin areas either don't use JS or are not JS heavy and can probably just use rack test here). I think there's good opportunity here for a reduction in test LOC while preserving/increasing the test coverage. We probably can't get to zero controller specs for non-API area, but we can probably get down to just edge cases or things that are really "about" the controllers as ruby classes and not as browser/user sessions.
- Add a selenium/headless configuration and add some system specs for some interaction which needs browser JS to work, but NOT the react app (yet). Get this configuration dialed in locally and on CI.
- Add something which needs the react app, but not (yet) the streaming server. Get that dialed locally/CI.
- Add something like what you have done here where the streaming server can be used as needed and hit by the react app being driven by a selenium/chromedriver/etc system spec.

I don't think that the larger api/controller request/system spec migrations need to be done in totality or before the last few bullets. That can be sort of a parallel effort. I do think that the last few bullets should be done in roughly that order.

I have this PR open -- https://github.com/mastodon/mastodon/pull/25027#issuecomment-1568558080 -- which I had planned as a first step along this path. My next thing after that's accepted would probably be migrating the feature specs to system specs, followed by adding something with headless browser, etc...

Separately - I know there is a PR open to split off the streaming server into it's own directory and more fully isolate the JS dependency management for the streaming server, the react app JS, and the non-node/streaming non-react JS libs the app uses. This is a great idea, and I think will make lots of things easier to iterate on if those deps are more cleanly isolated. It might make sense to complete that effort first as well, before this PR?

**Comment by renchap** [comment]
> > * I needed to set `LOCAL_DOMAIN` to `localhost` so the system tests work correctly. This need to be global, as the fixtures / DB use this value. I replaced the previous ngrok domain with `localhost`, but 1 non-system spec is failing and I dont know why
> 
> Out of curiosity, what was the driver of needing this change?

This is a good question, I reverted it and… it still works 🤔 I changed this very early when implementing this feature because it solved something, but I can not reproduce it anymore.

In any case, I would prefer to use an explicit example domain (as per RFC6761) rather than a ngrok domain (which was used at the time by Eugen when writing those initial specs to get real responses).

Regarding your roadmap, I would like system tests to be in place asap, as we are doing some heavy frontend refactor (touching major libs like the router) and it is very painful to test everywhere React components are rendered if nothing obvious broke (Share page, moderation screen…). Having basic system specs for those would at least help us being confident we did not miss anything obvious when refactoring.

I would prefer system specs to stay be always ran using a real browser, and use feature specs (which a few already exists) for non-JS end-to-end tests. I feel it is easier to have a specific kind of test for each Rack & Browser based end-to-end tests, rather than adding yet another spec attribute.

But then we need to figure out how we want to organise the end-to-end (system & feature) spec helpers. I am using `as_a_registered_user` from `ProfileStories`, but I feel this helper should be more generic and can be used in every end-to-end spec. I am not sure if the "stories" concept currently implemented is something idiomatic in rspec?

> Migrate existing controller specs to be system specs driving the browser

Then feature specs rather than system specs, so they use the Rack driver rather than the full-fledged Browser-driven one?

@ClearlyClaire told me that she does not want browser-based tests to be part of `rails spec`, so if we are to replace existing controller specs with higher-level ones, we still want them to be ran with `rails spec` and not require specific tooling.

> Separately - I know there is a PR open (#24702) to split off the streaming server into it's own directory and more fully isolate the JS dependency management for the streaming server, the react app JS, and the non-node/streaming non-react JS libs the app uses. This is a great idea, and I think will make lots of things easier to iterate on if those deps are more cleanly isolated. It might make sense to complete that effort first as well, before this PR?

The streaming split PR wont be merged before 4.2.0, and I would like end-to-end tests to be merge before it. But it wont change much, except the way we start the streaming server, so I dont think it is critical to have it merged before.

**Comment by mjankowski** [comment]
> Regarding your roadmap, I would like system tests to be in place asap, as we are doing some heavy frontend refactor (touching major libs like the router) and it is very painful to test everywhere React components are rendered if nothing obvious broke

Agreed, makes sense.

> I would prefer system specs to stay be always ran using a real browser, and use feature specs (which a few already exists) for non-JS end-to-end tests. I feel it is easier to have a specific kind of test for each Rack & Browser based end-to-end tests, rather than adding yet another spec attribute.

I think there's probably a small terminology disconnect here. Unfortunately the history and usage of "system" vs "end to end" vs "feature" tests/specs mean subtly different things to people both in general, but also in the narrow rspec/rails testing world.

The "feature specs" in rspec pre-date Rails having a built-in solution, but required you to do some manual setup on the capybara/transaction/database_cleaner side. In some rails release, "system tests" were included which did some of that setup for you automatically, and the rspec "system specs" are just a thin wrapper around this rails feature.

So with that history, I think at this point in time I use "Feature spec" and "System spec" almost interchangeably (in fact, I think you can basically copy one to the other and just rename what rspec `type` it is, and it will run just fine in most cases.

Putting the semantic and naming history aside, I agree with your general high level observations that:

- There's a pressing need to have something exercising the full stack, including the frontend JS and react app, because we are high risk of regressions there and the lack of coverage leads to slower dev cycles on anything touching those parts of the app.
- Ideally we can organize things in a way where each type of spec is organized code-wise with other things of that type that have the same running requirements/setup, rather than being interspersed and one-off tagged amongst other specs.
- We should have an intelligent default as to what runs locally in the basic `rake` or `rspec` run, and then the ability to run extra components either on CI or locally, as needed.

> But then we need to figure out how we want to organise the end-to-end (system & feature) spec helpers. I am using `as_a_registered_user` from `ProfileStories`, but I feel this helper should be more generic and can be used in every end-to-end spec.

Yes, it should be possible to do this and share helper methods across spec types. This is pretty easily configureable in the rspec setup.

> I am not sure if the "stories" concept currently implemented is something idiomatic in rspec?

This is definitely a style/approach to writing rspec feature/system specs (using "user stories" as guide to each spec example), but not a hard requirement of writing them if another style works better.


> > Migrate existing controller specs to be system specs driving the browser
> 
> Then feature specs rather than system specs, so they use the Rack driver rather than the full-fledged Browser-driven one?

Again maybe a naming disconnect, but yes, correct -- I think there's an improvement to be had on some of the controller specs where they'd gain from being run through the full request stack (which they would be as any of: request specs, feature specs, system specs but are not right now as controller specs), but where they don't actually need the full headless browser and/or JS setup, and could just run with the rack test driver.

> @ClearlyClaire told me that she does not want browser-based tests to be part of `rails spec`, so if we are to replace existing controller specs with higher-level ones, we still want them to be ran with `rails spec` and not require specific tooling.

Cool.

> In any case, I would prefer to use an explicit example domain (as per RFC6761) rather than a ngrok domain (which was used at the time by Eugen when writing those initial specs to get real responses).

I love following RFCs and would be delighted to see a cute little hostname like `host.example` kicking around. This is a fun RFC https://datatracker.ietf.org/doc/html/rfc2606#section-2

> The streaming split PR wont be merged before 4.2.0, and I would like end-to-end tests to be merge before it. But it wont change much, except the way we start the streaming server, so I dont think it is critical to have it merged before.

Fair enough. In the context of this PR, I think the bigger benefit is around the package management and having a narrower target of browser-loaded JS libs to think about, and not necessarily the streaming server startup process.

File: config/initializers/content_security_policy.rb (hunk 3 above)
Comment (renchap) [reply config/initializers/content_security_policy.rb:42]: This is caused by Webpack being in non-production mode in tests, and it uses `eval`.

Switching to `mode: "production"` for tests works.

**Comment by renchap** [comment]
Recent changes:
- Removed the `LOCAL_DOMAIN` changes, system tests work fine without it. But it would still be better to use a proper domain name for those.
- Removed `unsafe-eval` from CSP, as this was caused by Webpack not running in production mode
- Store logs and screenshots when the Github action fails

Things that need work:
- At the moment Webpacker is not ran before running system tests. Setting `compile: true` in `webpacker.yml` works, but it also compiles assets when running normal specs. We probably need to have it ran from rspec, before anything, if system tests are enabled for this run. And have an env variable or something to bypass it, for CI
- The streaming server needs to run on a random port (or at least a custom port), at the moment it collides with your local dev server
- I dont like having `as_a_logged_in_user` in `ProfileStories`, this needs to be moved somewhere more general, but I dont know how to organise it

**Comment by renchap** [comment]
> The "feature specs" in rspec pre-date Rails having a built-in solution, but required you to do some manual setup on the capybara/transaction/database_cleaner side. In some rails release, "system tests" were included which did some of that setup for you automatically, and the rspec "system specs" are just a thin wrapper around this rails feature.
> 
> So with that history, I think at this point in time I use "Feature spec" and "System spec" almost interchangeably (in fact, I think you can basically copy one to the other and just rename what rspec `type` it is, and it will run just fine in most cases.

Thanks for the history, I did not knew this.

I abused the difference between the 2 in my PR so feature specs a ran using the Rack driver, and system specs are ran using Capybara. But if this is not the way to do it, then I guess they can all be system specs, and we have an additional property to say if they use JS or not.

Or we can create our own category of tests to avoid doing this, depending if we want them all in the same directory (but specifying the use of JS for each suite), or a separate directory for each type.

**Comment by mjankowski** [comment]
> I abused the difference between the 2 in my PR so feature specs a ran using the Rack driver, and system specs are ran using Capybara. But if this is not the way to do it, then I guess they can all be system specs, and we have an additional property to say if they use JS or not.

Yeah, I'd say thats roughly best practice now -- make them all system specs and use tags per-example or per-file to control which driver is used in diff places. We can probably do that migration at same time that we migrate the profile stories helpers and make sure they are well named and accessible to whatever specs need them.

> Or we can create our own category of tests to avoid doing this, depending if we want them all in the same directory (but specifying the use of JS for each suite), or a separate directory for each type.

Putting all the things which need the streaming server spun up into one dir isn't strictly necessary ... but will absolutely make it easier to a) configure a default tag to apply to all specs in that dir, b) have the server started/stopped for those tags, c) isolated those specs from the default spec run so that they only run when opted into (env var, diff command, rake args, whatever).

File: spec/rails_helper.rb [hunk 4]
```diff
@@ -15,10 +15,15 @@
 Dir[Rails.root.join('spec', 'support', '**', '*.rb')].each { |f| require f }
 
 ActiveRecord::Migration.maintain_test_schema!
-WebMock.disable_net_connect!(allow: Chewy.settings[:host])
+WebMock.disable_net_connect!(allow: Chewy.settings[:host], allow_localhost: true)
```
[the remark this answers is not in the document]
Comment (renchap) [reply spec/rails_helper.rb]: I tried to configure them in `before :all` spec hooks, but Capybara does requests to the webdriver server outside of those (when shutting down the browser) so I enabled the setting globally, but only when system tests are enabled.

**Comment by renchap** [comment]
All comments above have been addressed, except the part about how the specs are organised (system vs feature) and how the helpers are organised (it still reuses the existing stories helpers).

Marking it as ready for review, I feel it can be merged as-it-is, or it needs somebody with more spec-cred (heh. I am tired.) to work on those aspects.

File: spec/rails_helper.rb [hunk 5]
```diff
@@ -1,6 +1,14 @@
 # frozen_string_literal: true
 
 ENV['RAILS_ENV'] ||= 'test'
+
+# This needs to be defined before Rails is initialized
+RUN_SYSTEM_SPECS = ENV.fetch('RUN_SYSTEM_SPECS', false)
+
+if RUN_SYSTEM_SPECS
+  STREAMING_PORT = rand(4001..4999)
```
Comment (mjankowski) [inline spec/rails_helper.rb:9]: It seems sort of odd to have a random assignment like this for an env that in theory we fully control (the CI setup) ... whats the background here?

File: spec/rails_helper.rb [hunk 6]
```diff
@@ -56,6 +68,8 @@ def get(path, headers: nil, sign_with: nil, **args)
 end
 
 RSpec.configure do |config|
+  # This is set before running spec:system, see lib/tasks/tests.rake
+  config.filter_run_excluding type: :system unless RUN_SYSTEM_SPECS
```
Comment (mjankowski) [inline spec/rails_helper.rb:72]: This is a good approach for this PR, which introduces this type.

If/when we do more migration/shuffling of the system/feature specs, we could migrate this to a specific tag/dir instead of all the system specs.

File: spec/rails_helper.rb [hunk 7]
```diff
@@ -83,8 +97,9 @@ def get(path, headers: nil, sign_with: nil, **args)
   end
 
   config.before :each, type: :feature do
-    https = ENV['LOCAL_HTTPS'] == 'true'
-    Capybara.app_host = "http#{https ? 's' : ''}://#{ENV.fetch('LOCAL_DOMAIN')}"
+    # https = ENV['LOCAL_HTTPS'] == 'true'
```
Comment (mjankowski) [inline spec/rails_helper.rb:100]: Commented lines left by accident?

File: spec/rails_helper.rb [hunk 8]
```diff
@@ -95,6 +110,35 @@ def get(path, headers: nil, sign_with: nil, **args)
     stub_jsonld_contexts!
   end
 
+  config.before :suite do
+    if RUN_SYSTEM_SPECS
```
Comment (mjankowski) [inline spec/rails_helper.rb:114]: I think instead of the env var conditional here, you could do something like `config.before :suite, type: :system`, which we could eventually migrate to a tag if need.

File: spec/rails_helper.rb [hunk 9]
```diff
@@ -95,6 +110,35 @@ def get(path, headers: nil, sign_with: nil, **args)
...
+      Webpacker.compile
+      streaming_server_manager.start(port: STREAMING_PORT)
+    end
+  end
+
+  config.after :suite do
```
Comment (mjankowski) [inline spec/rails_helper.rb:120]: The `before` had a conditional, but this does not?

File: spec/spec_helper.rb [hunk 10]
```diff
@@ -52,3 +52,80 @@ def expect_push_bulk_to_match(klass, matcher)
     'args' => matcher,
   }))
 end
+
+class StreamingServerManager
```
Comment (mjankowski) [inline spec/spec_helper.rb:56]: Can this file live somewhere other than within spec_helper?

File: spec/spec_helper.rb [hunk 11]
```diff
@@ -52,3 +52,80 @@ def expect_push_bulk_to_match(klass, matcher)
...
+    @running_thread = Thread.new do
+      Open3.popen2e(
+        {
+          'REDIS_NAMESPACE' => ENV.fetch('REDIS_NAMESPACE'),
+          'DB_NAME' => "#{ENV.fetch('DB_NAME', 'mastodon')}_test#{ENV.fetch('TEST_ENV_NUMBER', '')}",
+          'RAILS_ENV' => ENV.fetch('RAILS_ENV', 'test'),
+          'NODE_ENV' => ENV.fetch('NODE_ENV', 'test'),
+          'PORT' => port.to_s,
+        },
+        'node index.js', # must not call yarn here, otherwise it will fail because yarn does not send signals to its child process
```
Comment (mjankowski) [inline spec/spec_helper.rb:79]: Is the main reason this class needs to exist so that the rspec run can manage the start/stop of the streaming server appropriately?

I'm assuming it would be straightforward on CI to manage it more directly, but that needing it to start/stop within the spec suite is why this approach was chosen?

File: spec/support/stories/profile_stories.rb [hunk 12]
```diff
@@ -9,6 +9,8 @@ def as_a_registered_user
       email: email, password: password, confirmed_at: confirmed_at,
       account: Fabricate(:account, username: 'bob')
     )
+
+    Web::Setting.where(user: bob).first_or_initialize(user: bob).update!(data: { introductionVersion: 201812160442020 }) if finished_onboarding # rubocop:disable Style/NumericLiterals
```
Comment (mjankowski) [inline spec/support/stories/profile_stories.rb:13]: Is this related to the PR?

**Comment by mjankowski** [comment]
> All comments above have been addressed, except the part about how the specs are organised (system vs feature) and how the helpers are organised (it still reuses the existing stories helpers).

That stuff can all be refactored as we actually make those changes and doesnt need to be perfect now.

> Marking it as ready for review, I feel it can be merged as-it-is, or it needs somebody with more spec-cred (heh. I am tired.) to work on those aspects.

I hadnt looked at this in detail before - just left a few comments.

File: spec/spec_helper.rb (hunk 11 above)
Comment (renchap) [reply spec/spec_helper.rb:79]: This is not only for CI, but you need to run the streaming server, against the correct DB, when running the test suite locally as well. I found it to be easier to automatically handle this rather than writing docs & checks to ensure it has been started correctly by the user

File: spec/support/stories/profile_stories.rb (hunk 12 above)
Comment (renchap) [reply spec/support/stories/profile_stories.rb:13]: Yes, I needed a way to create users who do not have the onboarding screen for the 2 specs I wrote

File: spec/spec_helper.rb (hunk 10 above)
Comment (renchap) [reply spec/spec_helper.rb:56]: Yes, but I cas not sure where to put it. I can move it somewhere if you have a better idea!

File: spec/rails_helper.rb (hunk 8 above)
Comment (renchap) [reply spec/rails_helper.rb:114]: Unfortunately this does not work, `before :suite` does not support tag filtering. It kind of makes sense as this is launched before the whole suite, not before a specific spec or group.

File: spec/rails_helper.rb (hunk 9 above)
Comment (renchap) [reply spec/rails_helper.rb:120]: If the streaming server is not started then this is a no-op, so I chose to not add a conditional here

File: spec/rails_helper.rb (hunk 5 above)
Comment (renchap) [reply spec/rails_helper.rb]: Change it an a specific env variable, with a default.

The main issue was that re-using the port 4000 was conflicting with any local dev environment you might be running, so I figured that making it random would solve this and allow parallel system tests if needed as some point.

But it makes more sense to have it set by the environment, so here we go

File: config/webpack/tests.js [hunk 13]
```diff
@@ -5,5 +5,5 @@ const { merge } = require('webpack-merge');
 const sharedConfig = require('./shared');
 
 module.exports = merge(sharedConfig, {
-  mode: 'development',
+  mode: 'production',
```
Comment (ClearlyClaire) [inline config/webpack/tests.js:8]: What does this change exactly?

**Review by ClearlyClaire — APPROVED** [review]

File: config/webpack/tests.js (hunk 13 above)
Comment (renchap) [reply config/webpack/tests.js:8]: In `development`, Webpack uses `eval` to handle dynamic imports and other features, which requires a CSP change

Comment (ClearlyClaire) [reply config/webpack/tests.js:8]: Oh, right, that makes sense!
